### ISSUE TITLE
Query: compilation error for queries with groupby anonymous object and table splitting

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#11831]
Repro:

```cs
    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();
            }

            using (var ctx = new MyContext())
            {
                var query = (from f in ctx.Foos
                              join b in ctx.Bars on f.Id equals b.Id into grouping
                              from b in grouping.DefaultIfEmpty()
                              join bz in ctx.Bazs on f.Id equals bz.Id into grouping2
                              from bz in grouping2.DefaultIfEmpty()

                              select new { bname = b.Name, bzname = bz.Name }).GroupBy(g => new { g.bname, g.bzname }).Select(g => new { MyKey = g.Key.bname, cnt = g.Count() + 5 }).ToList();
            }
        }
    }

    public class MyContext : DbContext
    {
        public DbSet<Foo> Foos { get; set; }
        public DbSet<Bar> Bars { get; set; }
        public DbSet<Baz> Bazs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=.;Database=TableSplittin;Trusted_Connection=True;MultipleActiveResultSets=True");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Foo>().ToTable("Fubarbaz");
            modelBuilder.Entity<Bar>().ToTable("Fubarbaz");
            modelBuilder.Entity<Baz>().ToTable("Fubarbaz");

            modelBuilder.Entity<Foo>().HasOne<Bar>().WithOne().HasForeignKey<Bar>(b => b.Id);
            modelBuilder.Entity<Foo>().HasOne<Baz>().WithOne().HasForeignKey<Baz>(b => b.Id);
        }
    }

    public class Foo
    {
        public int Id { get; set; }

        public string Name { get; set; }
    }

    public class Bar
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class Baz
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }
```

exception:

```
'The binary operator NotEqual is not defined for the types 'Microsoft.EntityFrameworkCore.Storage.ValueBuffer' and 'Microsoft.EntityFrameworkCore.Storage.ValueBuffer'.'

   at System.Linq.Expressions.Expression.GetEqualityComparisonOperator(ExpressionType binaryType, String opName, Expression left, Expression right, Boolean liftToNull)
   at System.Linq.Expressions.Expression.NotEqual(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
   at System.Linq.Expressions.Expression.NotEqual(Expression left, Expression right)
   at Microsoft.EntityFrameworkCore.Query.Expressions.Internal.NullConditionalExpression.Reduce() in D:\git\EntityFrameworkCore\src\EFCore\Query\Expressions\Internal\NullConditionalExpression.cs:line 93
   at System.Linq.Expressions.Expression.ReduceAndCheck()
   at System.Linq.Expressions.Expression.ReduceExtensions()
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExtensionExpression(Expression expr, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpression(Expression node, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.ChildRewriter.Add(Expression expression)
   at System.Linq.Expressions.Compiler.StackSpiller.ChildRewriter.AddArguments(IArgumentProvider expressions)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteNewExpression(Expression expr, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpression(Expression node, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteConditionalExpression(Expression expr, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpression(Expression node, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteMemberExpression(Expression expr, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpression(Expression node, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.ChildRewriter.Add(Expression expression)
   at System.Linq.Expressions.Compiler.StackSpiller.ChildRewriter.AddArguments(IArgumentProvider expressions)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteNewExpression(Expression expr, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpression(Expression node, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.RewriteExpressionFreeTemps(Expression expression, Stack stack)
   at System.Linq.Expressions.Compiler.StackSpiller.Rewrite[T](Expression`1 lambda)
   at System.Linq.Expressions.Expression`1.Accept(StackSpiller spiller)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda)
   at System.Linq.Expressions.LambdaExpression.Compile()
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ProjectionShaper.Create(Shaper originalShaper, LambdaExpression materializer) in D:\git\EntityFrameworkCore\src\EFCore.Relational\Query\ExpressionVisitors\Internal\ProjectionShaper.cs:line 37
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore.Relational\Query\RelationalQueryModelVisitor.cs:line 1213
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs:line 910
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore.Relational\Query\RelationalQueryModelVisitor.cs:line 379
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs:line 175
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel) in D:\git\EntityFrameworkCore\src\EFCore\Storage\Database.cs:line 70

```

QM:

```
from IGrouping<<>f__AnonymousType3<string, string>, <>f__AnonymousType3<string, string>> g in 
    (from Foo f in DbSet<Foo>
    join Bar b in DbSet<Bar>
    on [f].Id equals [b].Id into grouping
    from Bar b in 
        (from Bar <generated>_1 in [grouping]
        select [<generated>_1]).DefaultIfEmpty()
    join Baz bz in DbSet<Baz>
    on [f].Id equals [bz].Id into grouping2
    from Baz bz in 
        (from Baz <generated>_1 in [grouping2]
        select [<generated>_1]).DefaultIfEmpty()
    select new <>f__AnonymousType3<string, string>(
        [b]?.Name, 
        [bz]?.Name
    )).GroupBy(new <>f__AnonymousType3<string, string>(
            [b]?.Name, 
            [bz]?.Name
        ), new <>f__AnonymousType3<string, string>(
            [b]?.Name, 
            [bz]?.Name
        ))
select new <>f__AnonymousType4<string, int>(
    [g]?.Key.bname, 

        (from <>f__AnonymousType3<string, string> <generated>_1 in [g]
        select [<generated>_1]).Count() + 5
)
```

materializer that throws the exception:

```
(QueryContext queryContext | ValueBuffer g) => new <>f__AnonymousType4<string, int>(
    !(g.IsEmpty) ? new <>f__AnonymousType3<string, string>(
        g?.Name, 
        g?.Name
    ) : default(<>f__AnonymousType3<string, string>).bname, 
    (int)Nullable<int> TryReadValue(g, 0, null)
)
```

**Comment by smitpatel** [comment]
The issue is this line.
> [g]?.Key.bname, 

We re-map [g].Key in translator to bind. But in this case it is not `MemberExpression`.
`NullConditionalExpression` here seems incorrect. Groups are never null. They can have null key and zero or more elements.

**Comment by smitpatel** [comment]
Issue is in `NavigationRewritingExpressionVisitor`
`NeedsNullCompensation` method should account for GroupResultOperator instead of inspecting whole QM for just `DefaultIfEmpty`

**Comment by maumar** [comment]
fixed in 2ba844ec5f877e57a879e8d88ad7df14b2c128f4
